### ISSUE TITLE
[question] package_id() to include requirements outside conan

### ISSUE THREAD

**Issue by iiknd** [conan-io/conan#7881]
Hi,

what is the recommended way to include "external" dependencies into the package_id calculation?
I have a conan package that depends on libraries outside the "conan realm", those are installed separately to the system.

If the version of the external libraries change it may cause binary break of the conan package(s) thus I'd want to take this into account in "def package_id(self)".

The documentation says that conan hashes the self.info object to generate the binary ID.

Is it possible add custom fields into the self.info object which would be then taken into account?

Dummy example:

```
def package_id(self):
    self.info.external_requires.lib_name = "foobar"
    self.info.external_requires.lib_ver = "1.2.3"
    self.info.external_requires.lib_os = "linux"
    ...
```

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @unzap 

There are a couple of approaches that you can follow:

- You can create a ``foobar/system`` package recipe that doesn't actually contain binaries, it is just a proxy for the system one. It can also manage to install the system one with ``system_requirements`` calling the system package manager. This is a nice way to control conflicts that could appear, for example if some other package in the dependencies try to use a ``foobar/1.3`` and another one tries to use the system one, it will give you a conflict. This also gives a nice way to centralize possible definition of include paths, preprocessor definitions, etc. We use this approach in ConanCenter for xorg, opengl and other system things.

- You could add some settings into the default settings.yml, that account for this system installation. Some users have directly model something like ``foobar_platform: version: [1, 2, 3, 4...]`` into the settings.yml. It can be abstracted to represent also a set of system dependencies, not just one by one. While this is doable, I think the above approach is a bit more correct.

Please let me know if this helps.

**Comment by iiknd** [comment]
Hi @memsharded and thanks for the tip!

Am I looking the correct file here: https://github.com/conan-io/conan-center-index/blob/master/recipes/xorg/all/conanfile.py ?


So if I want to proxy a system lib per system lib version I could have something like this?

```
class SystemLib(ConanFile):
    name = "theSystemLib"

    def package_id(self):
        self.info.header_only()  # do I need this?

        self.version = # check the system lib + version from the system env and set accordingly
```

EDIT:

I found this also: https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#using-package-id-for-package-dependencies

`self.info.requires["my_other_lib"] is a RequirementInfo object.`

I was looking if it was possible to instantiate a RequirementInfo object in the recipe and fill that with info from the system lib.
But that class seems to be internal.

**Comment by memsharded** [comment]
Hi @unzap 

Yes https://github.com/conan-io/conan-center-index/blob/master/recipes/xorg/all/conanfile.py is one of them.

Please, don't use uppercase in package names. We strongly recommend using always lowercase in package names.

You don't really need the ``self.info.header_only()`` unless the recipe is defining settings, options or has transitive requirements. In the Xorg example, the recipe has a constraint to "only Linux", and then needs to declare the ``os`` setting, in order to be able to get the input to raise an error, and then later it is removing that setting (strictly it won't be necessary, as you would have only 1 "package" anyway, as it only allows 1 value=Linux)

You cannot set the ``version`` at that point. The "version" of the package is defined way before, when it is created. The same "proxy" recipe is valid for different underlying "system" recipes, so this is why we name those packages ``xorg/system``, ``opengl/system``.

If you want to have different versions for your proxy recipe, you would need:
- Use the ``set_version()`` method, to detect the system version and assign it dynamically, at package creation time (export/create)
- If you want your consumers to implement different binaries for different versions of your proxy recipe, make sure to adapt (if semver is not good enough), like:
```python
def package_id(self):
    self.info.requires["mysystemdep"].full_version_mode()
```

Please let me know if this helps, thanks!

**Comment by iiknd** [comment]
I did a quick test and I can get a different package (e.g. foobar) in the cache per proxy version. Thanks for the help!

But I did run into another issue when there are multiple versions present of the system lib at the same time.
I was able to get that to work for the "conan install" part. I can control the system lib version e.g. via env:

`$conan install foobar/<version>@user/channel ... --build=missing -e USE_SYSTEM_LIB_VERSION=1.1`

Then in the foobar's conanfile.py I can dynamically set the requirement:

```
class FooBar(ConanFile):
    def requirements(self):
        try:
            self.requires("systemlibproxy/{0}@user/channel".format(os.environ['USE_SYSTEM_LIB_VERSION']))
        except KeyError:
            # not much we can do e.g. when calling "conan info" as the 'USE_SYSTEM_LIB_VERSION' env probably is not present
            pass

    def package_id(self):
        self.info.requires["systemlibproxy"].full_package_mode()
```

But this obviously won't work when you call e.g. "conan info" for the foobar package.
It tries to execute the requirements() which can not set the requirement dynamically.
Consequently the packge_id() will fail also.

I could check if  self.info.requires["systemlibproxy"] actually exists in package_id(), but then at least when calling "conan info" the package_id() would produce wrong result?



This is the implementation of the proxy:

```
class SystemLibProxy(ConanFile):
    name = "systemlibproxy"
    settings = "os", "compiler", "build_type", "arch"

    def source(self):
        pass

    def build(self):
        pass

    def package(self):
        pass

    def package_id(self):
        pass

    def package_info(self):
        pass
```

```
$conan export . systemlibproxy/1.0@user/channel
$conan export . systemlibproxy/1.1@user/channel
$conan export . systemlibproxy/1.2@user/channel
```

EDIT:

conan info does accept -e or --profile as arguments, i.e. those pull-in the env which would be used during the actual build.
So is it a valid requirement for the consumers to pass -e/--profile (with [env]) when calling "conan info" for this package?

**Comment by memsharded** [comment]
I think ``self.requires("systemlibproxy/{0}@user/channel".format(os.environ['USE_SYSTEM_LIB_VERSION']))`` is probably not a very good idea. You are implementing a "conditional requirement" in a way that is not known to Conan.

If your proxy recipe is completely empty, then using settings might be a better and simpler way. I understood that you really needed to pass information of the system lib to the consumers, and then your ``systemlibproxy`` ``package_info()`` method would be populated with information. If this is not the case, then maybe the best would be to just add to your settings.yml file (assuming the system lib is in Linux):

```yaml
os:
    Linux:
           mysystemlib: [None, 1.0, 1.1, 1.2]
```
That is all. No need to change recipes, or define requirements, just define the system lib version in your profiles, and you will get different binaries for all recipes that declare the ``os`` setting.

Please let me know if this makes sense.

**Comment by iiknd** [comment]
If I publish such conan package then other people would need to adapt their settings.yml file accordingly if they want to include the "foobar" in their project?
Besides that it looks like a very clean approach indeed. I need to think about this..

**Comment by memsharded** [comment]
> If I publish such conan package then other people would need to adapt their settings.yml file accordingly if they want to include the "foobar" in their project?

Yes, but you have the ``conan config install`` feature for this, it is very recommended in production for teams. You can use it not only for the settings.yml, but also for your profiles, your remotes definitions, your conf (including required Conan version)...

When you mean "publish", the other people are people from your team, or is it something open source open to the world?

**Comment by iiknd** [comment]
> When you mean "publish", the other people are people from your team, or is it something open source open to the world?

Yes, if I wanted to publish this package to e.g. conan-center or similar.

This is actually now more a generic question.
If a development team wants to utilize a package A e.g. from conan-center, but the recipe of A requires a custom settings.yml file.
The development team needs to merge the settings for A into their settings.yml file?

**Comment by memsharded** [comment]
> Yes, if I wanted to publish this package to e.g. conan-center or similar.

Yes, packages with custom settings cannot be published to ConanCenter. But the trick with the env-var will not work either, it is not possible to do such a thing. We would need to understand what system library you are depending and modeling and think about possible alternatives suited for ConanCenter wide audience.

> If a development team wants to utilize a package A e.g. from conan-center, but the recipe of A requires a custom settings.yml file.
The development team needs to merge the settings for A into their settings.yml file?

You cannot have a package A in ConanCenter that requires custom settings. You can put it in your own Artifactory, and your users would need to add that remote, and potentially ``conan config install`` your settings in order to be able to consume your packages.

Custom settings can also work and reuse packages from ConanCenter as long as they are appending new values, or appending settings or subsettings that do not affect to binaries in ConanCenter. Typically the maintainers will keep updated their own custom settings.yml integrating new changes that might come from the upstream conan one.

**Comment by iiknd** [comment]
> I think self.requires("systemlibproxy/{0}@user/channel".format(os.environ['USE_SYSTEM_LIB_VERSION'])) is probably not a very good idea. You are implementing a "conditional requirement" in a way that is not known to Conan.

About the conditional requirements, the documentation has an example: https://docs.conan.io/en/latest/reference/conanfile/methods.html#requirements

```
def requirements(self):
    if self.options.myoption:
        self.requires("zlib/1.2@drl/testing")
    else:
        self.requires("opencv/2.2@drl/stable")
```

Isn't this build time resolution as well, compared to my example of the dynamic version number? If you run "conan info" for the example package above it would fail if one does not pass correct options/profile for "conan info -o ..."?

**Comment by memsharded** [comment]
The key with ``if self.options.myoption:`` is:

- Options are always explicitly modeled in the recipes. 
- They need to have a value assigned, if a value is not given it will error
- They can have explicit default values in recipes.
- They can be listed with commands like inspect or in server WebUI like Artifactory
- The options are explicitly modeled in the ``package_id()``, changing an option change the resulting ``package_id``.
- Options have a range of accepted values. Typos or incorrect values will also raise errors.

In summary, options are way more controlled and explicit modeled than environment variables. In any case, the potential anti-pattern that I see is defining requirements version in such a way, instead of using other mechanisms like overriding requirements.

**Comment by iiknd** [comment]
Hi, thanks for the comments!

So to re-model the proxy to be "versionless" like this:


```
class SystemLibProxy(ConanFile):
    name = "systemlibproxy" 
    version = "1.0"  # can be anything
    settings = "os", "compiler", "build_type", "arch"
    options = {"use_system_lib": ["ANY"]}
```


Then one can:

`conan install ... -o use_system_lib=1.2.3`

Would this approach work regarding the unique package_id?

One more question about conditional requirements. From documentation:

```
def requirements(self):
    if self.options.myoption:
        self.requires("zlib/1.2@drl/testing")
    else:
        self.requires("opencv/2.2@drl/stable")
```

If the "system lib" would become a real conan package some day the above looks like a way to
make this conditional in the recipe?

```
def requirements(self):
    if self.options.use_system_lib:
        self.requires("systemlibproxy/1.0@user/channel")
    else:
        self.requires("the_system_lib_from_eg_conan_center/1.2.3@user/channel")
```

**Comment by memsharded** [comment]
> Would this approach work regarding the unique package_id?

Yes, options of the dependencies are hashed in the package_id, so that will produce a different binary of the consumers for different values of the options.

> If the "system lib" would become a real conan package some day the above looks like a way to
make this conditional in the recipe?

Yes, you might use ``if self.options.use_system_lib`` as a condition to switch between a system package and a Conan provided one.

**Comment by iiknd** [comment]
@memsharded  thanks a lot for the help! This approach works for me :)

**Comment by jgsogo** [comment]
Wow! Here there is a very interesting topic. When we introduced `system` packages we thought it was no-sense to add versioning to those packages because you need to use whatever was installed in the system (or chosen by the underlying package manager), with no way to change those dependencies. With this in mind, the best approach for consumers was to write `xorg/system` as a requirement and use whatever was provided by the OS/distro to compile and link (like the `glibc` problem).


Is this a real scenario keeping in mind only actual system packages? **These packages are always dynamically linked**, and they are ABI compatible (at least forward compatible), any breaking change would be as problematic as the C++ `std::string` ABI break, other libraries and apps in the system are using them. **IMHO, for system packages, any ABI breakage should be encoded at the level of `os/distro` settings** because these packages interact which each other in the system. In the same way you couldn't change the `libstdc++.so` ABI without changing all depending packages (and apps) in the system.

---

I think that using the approach we did for `system` packages for other 3rd party requirements that are not _conanized_ can be confusing. This is a different problem, these packages are not system packages, probably they are not even _installed_ at the system level (otherwise we couldn't install another version breaking the ABI without upgrading many other libraries and apps), they are just provided by other means to be used and linked.

For these packages, I think that they should totally use versioning, `foobar/v1.2.3@user/channel`, some kind of _proxy_ recipe (while the package is conanized) for enterprise/personal usage. In ConanCenter we are not accepting `system` packages unless they are really needed (very exceptional cases) or they are actual _system_ packages that cannot be changed in the OS.

Maybe knowing more about the specific library could help.

**Comment by iiknd** [comment]
Hi,

some examples of system or other libraries on the system which are installed outside Conan:
- Vulkan
- Qt
- Tck/Tk
- etc.

I.e. one may not always take the latest available version.

**Comment by jgsogo** [comment]
For these regular packages, I'd recommend using regular references like `vulkan/1.2.140@unzap/my-pkg-manager`, and you can write any Python code in your recipe: it can download the headers and binaries from custom your location and place them in your system (instead of the clone+build of common recipes), and populate the `cpp_info` attribute information in the `package_id()` method.

Using this approach, the integration of those packages in your Conan graph is totally transparent: you can change versions, Conan will resolve conflicts,... and eventually you only need to evolve the recipe or change the requirement to start using the `vulkan` package from ConanCenter once it is available.

**Comment by memsharded** [comment]
@unzap an important distinction of "system libraries" and libraries that are installed outside of Conan, as we name them in ConanCenter:

- "system library" is something that heavily belongs to the system, like X11. You can typically only have 1 version of it, typically managed at the OS installation level, part of the distro. You probably cannot build an app with a different version of X11 and run it in your system. Drivers, and low level things like OpenGL drivers (not high level libs), could belong to this. It is very difficult that Conan will be building, packaging and installing X11 at the user level (remember Conan works at the user level)
- Libraries like Qt are "regular" libraries. You can have different versions of it "installed" in your system, and you can easily have different projects using different versions and everything is working.

For the first type you might use a ``x11/system`` Conan package to proxy it, because the internal version is almost transparent. Your binaries will have the binary compatibility of the system, and if anything, you might need to model better the OS distro for the binary compatibility, but not explicitly the x11 system.

For the second one, using the real version like ``qt/1.3.4`` or ``vulkan/3.4.5`` is probably better, even if they are just "proxy" recipes to the library installed by other means, because it will explicitly model your dependency graph, handle conflicts, can automate checks and different tasks and definitions of information for consumers, and also be prepared if at some point you even want to fully manage those binaries within Conan. Plus it will naturally model your packages binary compatibility.

**Comment by memsharded** [comment]
So at the moment, it seems the above described approaches would be the best solutions: 
- ``pkg/system`` for system level libraries that can only have 1 version which is in practice bound to the OS/distro, 
-  ``pkg/version``, proxy recipe, that can model and use different versions of the same library that is not installed by Conan, but that in theory could be converted to a Conan package.

Both require relatively low effort, explicitly model the dependencies to the necessary level, bringing benefits like conflict detection, wrapping and encapsulation of information in a single place, etc.

Adding extra arbitrary information to the ``package_id`` doesn't seem a good approach so far. 
I think we can close this issue now, but please @unzap let us know how it goes and if you have any further question don't hesitate to comment or open new issues. Thanks!
